File: typescript/src/structural/proxy/proxy/Document.ts
export interface Document{
    display: (user:string) => void;
}


export class RealDocument implements Document {
    private filename: string
    private content: null
    
    constructor(filename: string, content: null){
        this.filename = filename
        this.content = content
    }
    
    sleep(ms: number): Promise<void> {
        return new Promise(resolver => setTimeout(resolver, ms))
    }
    
    async loadFromDisk(): Promise<void> {
        console.log(`📄 Load ${this.filename} document from disk...`)
        await this.sleep(3000)
    }

    async display(user: string): Promise<void> {
        if (!this.content) {
            await this.loadFromDisk(); // Esperar a que el documento se cargue
        }
        console.log(`${user} is watching document: ${this.filename}`);
    }
}

File: python/structural/proxy/proxy/document.py
from abc import ABC, abstractmethod
import time

class Document(ABC):
    @abstractmethod
    def display(self, user: str):
        pass


class RealDocument(Document):
    def __init__(self, filename: str):
        self.filename = filename
        self.content = None

    def load_from_disk(self):
        print(f"📄 Load {self.filename} document from disk...")
        time.sleep(3)
        self.content = f"Contenido {self.content}"

    def display(self, user: str):
        if not self.content:
            self.load_from_disk()
        print(f"{user} is watching document: {self.filename}")
